lib/SILGen/SILGenExpr.cpp

```diff
@@ -3532,6 +3537,70 @@ SILGenModule::emitKeyPathComponentForDecl(SILLocation loc,
   llvm_unreachable("unknown kind of storage");
 }
 
+// There is some uglyness here because this function has to not only handle
+// the generation of default arguments but it also has to create a new TupleExpr
+// or ParenExpr depending on how many default args are found.
+void RValueEmitter::generateDefaultArgsForKeyPathSubscript(
+    SubscriptDecl *S, KeyPathExpr::Component &component) {
+  // Collect both all the args (default and not) into one place and
+  // keep track of their types.
+  llvm::SmallVector<TupleTypeElt, 4> newTypeEls;
+  newTypeEls.reserve(S->getIndices()->size());
+  llvm::SmallVector<Expr *, 4> newArgs;
+  newArgs.reserve(S->getIndices()->size());
+  size_t index = 0;
+  for (auto param : *S->getIndices()) {
+    newTypeEls.push_back(TupleTypeElt(param->getType()));
+
+    if (param->isDefaultArgument()) {
+      newArgs.push_back(param->getDefaultValue());
+    } else {
+      if (auto paren = dyn_cast<ParenExpr>(component.getIndexExpr())) {
+        newArgs.push_back(paren->getSubExpr());
+      } else if (auto tupleExpr =
+                     dyn_cast<TupleExpr>(component.getIndexExpr())) {
+        newArgs.push_back(tupleExpr->getElement(index));
+      }
+    }
+
+    (void)++index;
+  }
+
+  // This looks it's going to give us a TypleType but it will give us a
+  // ParenType when newTypeEls.size() == 1.
+  auto newIndexType = TupleType::get(newTypeEls, SGF.getASTContext());
+
+  // Unconditionally recreate the index expr based on the new type.
+  if (isa<ParenType>(newIndexType.getPointer())) {
+    assert(newArgs.size() == 1 && "If this was converted to a paren expr it "
+                                  "should have exact one element");
+    auto newIndexExpr = new (SGF.getASTContext())
+        ParenExpr(component.getIndexExpr()->getStartLoc(), newArgs[0],
+                  component.getIndexExpr()->getEndLoc(),
+                  /*hasTrailingClosure=*/false);
+
+    component.setIndexExpr(newIndexExpr);
+  } else if (isa<TupleType>(newIndexType.getPointer())) {
+    SmallVector<Identifier, 4> elementNames;
+    elementNames.reserve(newArgs.size());
+    if (auto tupleExpr = dyn_cast<TupleExpr>(component.getIndexExpr())) {
+      std::copy(tupleExpr->getElementNames().begin(),
+                tupleExpr->getElementNames().begin(), elementNames.begin());
+    } else {
+      std::transform(newArgs.begin(), newArgs.end(), elementNames.begin(),
+                     [](auto &&) { return Identifier(); });
+    }
+
+    auto newIndexExpr =
+        TupleExpr::createImplicit(SGF.getASTContext(), newArgs, elementNames);
+    component.setIndexExpr(newIndexExpr);
```

Comment:
This looks suspect to me - SILGen should be able to lower an AST without modifying it. If not, then it needs to be taught how to do it (or the AST you’re passing is not correct and should be changed in CSApply). However, I don’t think creating or modifying ASTs in the backend is the right thing to do.

Reply:
Yes, I think you're right. I'll see if I can find a way to do this differently.

Reply:
Aha! It's so simple. I've been looking at this code all day but I just saw exactly what needs to be changed. Update coming soon.